otwcode/otwarchive
# AO3-5522 Limit number of password resets

## Issue

https://otwarchive.atlassian.net/browse/AO3-5522

## Purpose

This adds a limit to the number of password resets that can be requested within a period of time. If the limit is reached, then reset requests will be prevented until the cooldown expires.

A new `users` table column is added: 
* `resets_requested`: The number of resets that have been requested. Integer, non-nullable, indexed, defaults to 0.

Two new archive configuration settings are added:
* `PASSWORD_RESET_LIMIT`: The limit to the number of resets within the `PASSWORD_RESET_COOLDOWN_HOURS` period. Defaults to 3.
* `PASSWORD_RESET_COOLDOWN_HOURS`: The period of time in which the reset request limit will be enforced. Defaults to 12.

## Testing Instructions

#### Test that password resets can be requested
* Open the login page (or login drop-down form)
* Click the "reset password" link
* Enter a valid, non-protected username or email address
* You should see a successful response and receive a password reset email

#### Test that password reset requests beyond the limit cannot be requested
* Follow the above instructions two more times
  * (or three times if it has been more than 12 hours since the last reset request)
* Each attempt should be successfull
* Follow the above instructions a fourth time
* You should see a response indicating that you cannot reset your password at this time. Response should indicate the date and time after which password resetting will again be available
* You should not receive a password reset email

#### Test that password resets can again be requested after waiting out the cooldown period
* Wait until after the time specified in the password reset error message
* Attempt to reset your password
* You should see a successful response and receive a password reset email

## Credit

What name and pronouns should we use to credit you in the [Archive of Our Own's Release Notes](https://archiveofourown.org/admin_posts?tag=1)?

Hunter Ada Smith (she/her)
